### ISSUE TITLE
RFC-5321 (§4.5.3.1.6) should be respected by MIMEText if possible

### ISSUE THREAD

**Issue by azmee-meta** [python/cpython#125867]
# Bug report

### Bug description:

When there is lots of 

```html
<tr>
	<td>36</td>
	<td>16</td>
	<td>-56%</td>
	<td></td>
	<td></td>
	<td></td>
	<td></td>
	<td></td>
	<td></td>
	<td></td>
	<td></td>
</tr>
<tr>
	<td></td>
</tr>
```

(i make this code dynamically, which sometimes lead to very big table/code, in that case the issue arise)

like this, then sometimes in `<td>` or `<tr>` tag render like this (`< td> < tr>`, put space unexpectedly) and the table breaks in the email body.

**Note:** i am using a little inline css in table:

```html
<html lang='en'>
<head><meta charset='UTF-8'></head>
<body><table border=1 bordercolor='black' cellspacing=0 cellpadding=5> <tr>
```

and

```py
msg.attach(MIMEText(html, 'html'))
```

### CPython versions tested on:

3.12

### Operating systems tested on:

Windows

**Comment by RanKKI** [comment]
Hi, can you please Include a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) ?

**Comment by azmee-meta** [comment]
> Hi, can you please Include a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) ?

when `html` variable renders like the following:

html="<html lang='en'><head><meta charset='UTF-8'></head><body><table border=1 bordercolor='black' cellspacing=0 cellpadding=5> <tr> <th>GroupName</th> <th>DS_ID</th> <th>DS_Name</th> <th>DS_MedExpectedSize</th> <th>DS_CurrentSize</th> <th>DS_PercentIncDec</th> <th>TG_ID</th> <th>TG_Name</th> <th>TG_MedExpectedSize</th> <th>TG_CurrentSize</th> <th>TG_PercentIncDec</th> <th>TG_Distinct_MedExpectedSize</th> <th>TG_Distinct_CurrentSize</th> <th>TG_Distinct_PercentIncDec</th></tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2453</td>  <td>TG - Country distribution - Operations</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2585</td>  <td>TG - Country distribution - Perfumery</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2587</td>  <td>TG - Country distribution - TTH</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  </tr>   <tr> <td>DSM</td>  <td>222</td>  <td>DSM users 2 - Department distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>EZ</td>  <td>353</td>  <td>EZK users</td>  <td>18</td>  <td>-3</td>  <td>-117%</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>KPN</td>  <td>696</td>  <td>KPN HR Servicedesk tickets</td>  <td>123</td>  <td>64</td>  <td>-48%</td>  <td>2512</td>  <td>KPN HR Servicedesk tickets</td>  <td>111</td>  <td>60</td>  <td>-46%</td>  <td>96</td>  <td>40</td>  <td>-58%</td>  </tr>   <tr> <td>NS</td>  <td></td>
   <td></td>  <td></td>  <td></td>  <td></td>  <td>2291</td>  <td>NS TG SSR - Email as Invitation Key - Prio 2</td>  <td>485</td>  <td>676</td>  <td>39%</td>  <td>258</td>  <td>408</td>  <td>58%</td>  </tr></table>    <br><br>   <div>Average expected DS/TG size on a Tuesday, will be calculated as the average TG sizes of the past 10 Tuesday.</div>   </body></html>"

`msg.attach(MIMEText(html, 'html'))`
`server.send_message(msg)`

this output comes:

`html="<html lang='en'><head><meta charset='UTF-8'></head><body><table border=1 bordercolor='black' cellspacing=0 cellpadding=5> <tr> <th>GroupName</th> <th>DS_ID</th> <th>DS_Name</th> <th>DS_MedExpectedSize</th> <th>DS_CurrentSize</th> <th>DS_PercentIncDec</th> <th>TG_ID</th> <th>TG_Name</th> <th>TG_MedExpectedSize</th> <th>TG_CurrentSize</th> <th>TG_PercentIncDec</th> <th>TG_Distinct_MedExpectedSize</th> <th>TG_Distinct_CurrentSize</th> <th>TG_Distinct_PercentIncDec</th></tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2453</td>  <td>TG - Country distribution - Operations</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2585</td>  <td>TG - Country distribution - Perfumery</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  <td>1289</td>  <
 td>2983</td>  <td>131%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2587</td>  <td>TG - Country distribution - TTH</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  </tr>   <tr> <td>DSM</td>  <td>222</td>  <td>DSM users 2 - Department distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>EZ</td>  <td>353</td>  <td>EZK users</td>  <td>18</td>  <td>-3</td>  <td>-117%</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>KPN</td>  <td>696</td>  <td>KPN HR Servicedesk tickets</td>  <td>123</td>  <td>64</td>  <td>-48%</td>  <td>2512</td>  <td>KPN HR Servicedesk tickets</td>  <td>111</td>  <td>60</td>  <td>-46%</td>  <td>96</td>  <td>40</td>  <td>-58%</td>  </tr>   <tr> <td>NS</td>  <td></td>
   <td></td>  <td></td>  <td></td>  <td></td>  <td>2291</td>  <td>NS TG SSR - Email as Invitation Key - Prio 2</td>  <td>485</td>  <td>676</td>  <td>39%</td>  <td>258</td>  <td>408</td>  <td>58%</td>  </tr></table>    <br><br>   <div>Average expected DS/TG size on a Tuesday, will be calculated as the average TG sizes of the past 10 Tuesday.</div>   </body></html>"`

**Comment by RanKKI** [comment]
I tried to reproduce this issue with

<details>
<summary>Client code</summary>

```python
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

html = """
<table border="1" cellspacing="0" cellpadding="5" role="table" tabindex="0">
   <tbody>
      <tr>
         <th>GroupName</th>
         <th>DS_ID</th>
         <th>DS_Name</th>
         <th>DS_MedExpectedSize</th>
         <th>DS_CurrentSize</th>
         <th>DS_PercentIncDec</th>
         <th>TG_ID</th>
         <th>TG_Name</th>
         <th>TG_MedExpectedSize</th>
         <th>TG_CurrentSize</th>
         <th>TG_PercentIncDec</th>
         <th>TG_Distinct_MedExpectedSize</th>
         <th>TG_Distinct_CurrentSize</th>
         <th>TG_Distinct_PercentIncDec</th>
      </tr>
      <tr>
         <td>DSM</td>
         <td>679</td>
         <td>DSM users 2 - Country distribution</td>
         <td>0</td>
         <td>142</td>
         <td>N/A</td>
         <td>2453</td>
         <td>TG - Country distribution - Operations</td>
         <td>2396</td>
         <td>3237</td>
         <td>35%</td>
         <td>2396</td>
         <td>3237</td>
         <td>35%</td>
      </tr>
      <tr>
         <td>DSM</td>
         <td>679</td>
         <td>DSM users 2 - Country distribution</td>
         <td>0</td>
         <td>142</td>
         <td>N/A</td>
         <td>2585</td>
         <td>TG - Country distribution - Perfumery</td>
         <td>1289</td>
         <td>2983</td>
         <td>131%</td>
         <td>1289</td>
         <td>2983</td>
         <td>131%</td>
      </tr>
      <tr>
         <td>DSM</td>
         <td>679</td>
         <td>DSM users 2 - Country distribution</td>
         <td>0</td>
         <td>142</td>
         <td>N/A</td>
         <td>2587</td>
         <td>TG - Country distribution - TTH</td>
         <td>1094</td>
         <td>152</td>
         <td>-86%</td>
         <td>1094</td>
         <td>152</td>
         <td>-86%</td>
      </tr>
      <tr>
         <td>DSM</td>
         <td>222</td>
         <td>DSM users 2 - Department distribution</td>
         <td>0</td>
         <td>142</td>
         <td>N/A</td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
      </tr>
      <tr>
         <td>EZ</td>
         <td>353</td>
         <td>EZK users</td>
         <td>18</td>
         <td>-3</td>
         <td>-117%</td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
      </tr>
      <tr>
         <td>KPN</td>
         <td>696</td>
         <td>KPN HR Servicedesk tickets</td>
         <td>123</td>
         <td>64</td>
         <td>-48%</td>
         <td>2512</td>
         <td>KPN HR Servicedesk tickets</td>
         <td>111</td>
         <td>60</td>
         <td>-46%</td>
         <td>96</td>
         <td>40</td>
         <td>-58%</td>
      </tr>
      <tr>
         <td>NS</td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td></td>
         <td>2291</td>
         <td>NS TG SSR - Email as Invitation Key - Prio 2</td>
         <td>485</td>
         <td>676</td>
         <td>39%</td>
         <td>258</td>
         <td>408</td>
         <td>58%</td>
      </tr>
   </tbody>
</table>
"""

msg = MIMEMultipart()
msg["From"] = "foo@example.com"
msg["To"] = "bar@example.com"
msg["Subject"] = "Test"
msg.attach(MIMEText(html, "html"))

with smtplib.SMTP("127.0.0.1", 1025) as server:
    server.send_message(msg)

```

</details>

<details>
<summary>Simple SMTP server</summary>

```python
import asyncio
from email.message import Message

from aiosmtpd.controller import Controller
from aiosmtpd.handlers import AsyncMessage


class PrintMessageHandler(AsyncMessage):
    async def handle_message(self, msg: Message):
        print(msg.as_bytes())
        return "OK"


async def main():
    handler = PrintMessageHandler()
    controller = Controller(handler, hostname="127.0.0.1", port=1025)
    controller.start()

    try:
        while True:
            await asyncio.sleep(3600)  # Keep the server running
    except (KeyboardInterrupt, asyncio.exceptions.CancelledError):
        controller.stop()


if __name__ == "__main__":
    asyncio.run(main())

```
</details>

The content printed looks good to me. Did I miss something?

```
b'Content-Type: multipart/mixed; boundary="===============8416919614510437434=="\nMIME-Version: 1.0\nFrom: foo@example.com\nTo: bar@example.com\nSubject: Test\nX-Peer: (\'127.0.0.1\', 60639)\nX-MailFrom: foo@example.com\nX-RcptTo: bar@example.com\n\n--===============8416919614510437434==\nContent-Type: text/html; charset="us-ascii"\nMIME-Version: 1.0\nContent-Transfer-Encoding: 7bit\n\n\n<table border="1" cellspacing="0" cellpadding="5" role="table" tabindex="0">\n   <tbody>\n      <tr>\n         <th>GroupName</th>\n         <th>DS_ID</th>\n         <th>DS_Name</th>\n         <th>DS_MedExpectedSize</th>\n         <th>DS_CurrentSize</th>\n         <th>DS_PercentIncDec</th>\n         <th>TG_ID</th>\n         <th>TG_Name</th>\n         <th>TG_MedExpectedSize</th>\n         <th>TG_CurrentSize</th>\n         <th>TG_PercentIncDec</th>\n         <th>TG_Distinct_MedExpectedSize</th>\n         <th>TG_Distinct_CurrentSize</th>\n         <th>TG_Distinct_PercentIncDec</th>\n      </tr>\n      <tr>\n         <td>DSM</td>\n         <td>679</td>\n         <td>DSM users 2 - Country distribution</td>\n         <td>0</td>\n         <td>142</td>\n         <td>N/A</td>\n         <td>2453</td>\n         <td>TG - Country distribution - Operations</td>\n         <td>2396</td>\n         <td>3237</td>\n         <td>35%</td>\n         <td>2396</td>\n         <td>3237</td>\n         <td>35%</td>\n      </tr>\n      <tr>\n         <td>DSM</td>\n         <td>679</td>\n         <td>DSM users 2 - Country distribution</td>\n         <td>0</td>\n         <td>142</td>\n         <td>N/A</td>\n         <td>2585</td>\n         <td>TG - Country distribution - Perfumery</td>\n         <td>1289</td>\n         <td>2983</td>\n         <td>131%</td>\n         <td>1289</td>\n         <td>2983</td>\n         <td>131%</td>\n      </tr>\n      <tr>\n         <td>DSM</td>\n         <td>679</td>\n         <td>DSM users 2 - Country distribution</td>\n         <td>0</td>\n         <td>142</td>\n         <td>N/A</td>\n         <td>2587</td>\n         <td>TG - Country distribution - TTH</td>\n         <td>1094</td>\n         <td>152</td>\n         <td>-86%</td>\n         <td>1094</td>\n         <td>152</td>\n         <td>-86%</td>\n      </tr>\n      <tr>\n         <td>DSM</td>\n         <td>222</td>\n         <td>DSM users 2 - Department distribution</td>\n         <td>0</td>\n         <td>142</td>\n         <td>N/A</td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n      </tr>\n      <tr>\n         <td>EZ</td>\n         <td>353</td>\n         <td>EZK users</td>\n         <td>18</td>\n         <td>-3</td>\n         <td>-117%</td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n      </tr>\n      <tr>\n         <td>KPN</td>\n         <td>696</td>\n         <td>KPN HR Servicedesk tickets</td>\n         <td>123</td>\n         <td>64</td>\n         <td>-48%</td>\n         <td>2512</td>\n         <td>KPN HR Servicedesk tickets</td>\n         <td>111</td>\n         <td>60</td>\n         <td>-46%</td>\n         <td>96</td>\n         <td>40</td>\n         <td>-58%</td>\n      </tr>\n      <tr>\n         <td>NS</td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td></td>\n         <td>2291</td>\n         <td>NS TG SSR - Email as Invitation Key - Prio 2</td>\n         <td>485</td>\n         <td>676</td>\n         <td>39%</td>\n         <td>258</td>\n         <td>408</td>\n         <td>58%</td>\n      </tr>\n   </tbody>\n</table>\n\n--===============8416919614510437434==--\n'
```

**Comment by azmee-meta** [comment]
**Please use the following code**

from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
import smtplib


msg = MIMEMultipart('alternative')
# msg = EmailMessage()
msg['From'] = 'abc@gmail.com'
msg['To'] = 'xyz@metatude.com'
    
msg['Subject'] = 'Unexpected DS/TG size found'

     # Create HTML content for email body
html = """<html lang='en'><head><meta charset='UTF-8'></head><body><table border=1 bordercolor='black' cellspacing=0 cellpadding=5> <tr> <th>GroupName</th> <th>DS_ID</th> <th>DS_Name</th> <th>DS_MedExpectedSize</th> <th>DS_CurrentSize</th> <th>DS_PercentIncDec</th> <th>TG_ID</th> <th>TG_Name</th> <th>TG_MedExpectedSize</th> <th>TG_CurrentSize</th> <th>TG_PercentIncDec</th> <th>TG_Distinct_MedExpectedSize</th> <th>TG_Distinct_CurrentSize</th> <th>TG_Distinct_PercentIncDec</th></tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2453</td>  <td>TG - Country distribution - Operations</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2585</td>  <td>TG - Country distribution - Perfumery</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2587</td>  <td>TG - Country distribution - TTH</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  </tr>   <tr> <td>DSM</td>  <td>222</td>  <td>DSM users 2 - Department distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>EZ</td>  <td>353</td>  <td>EZK users</td>  <td>18</td>  <td>-3</td>  <td>-117%</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>KPN</td>  <td>696</td>  <td>KPN HR Servicedesk tickets</td>  <td>123</td>  <td>64</td>  <td>-48%</td>  <td>2512</td>  <td>KPN HR Servicedesk tickets</td>  <td>111</td>  <td>60</td>  <td>-46%</td>  <td>96</td>  <td>40</td>  <td>-58%</td>  </tr>   <tr> <td>NS</td>  <td></td><td></td>  <td></td>  <td></td>  <td></td>  <td>2291</td>  <td>NS TG SSR - Email as Invitation Key - Prio 2</td>  <td>485</td>  <td>676</td>  <td>39%</td>  <td>258</td>  <td>408</td>  <td>58%</td>  </tr></table>    <br><br>   <div>Average expected DS/TG size on a Tuesday, will be calculated as the average TG sizes of the past 10 Tuesday.</div>   </body></html>"""
        # Attach HTML content to email
msg.attach(MIMEText(html, 'html'))
# msg.add_alternative(html, subtype='html')


with smtplib.SMTP('smtp.gmail.com', 587) as server:
    server.starttls()
    server.login('abc@gmail.com',
                    'xyz')
    server.send_message(msg)
    print("Email sent successfully.")

**Comment by picnixz** [comment]
Can you cleanly show what you would expect as an output (e.g., using a screenshot) and highlight the differences? the table formatting on GitHub makes it hard to follow.

**Comment by azmee-meta** [comment]
![image](https://github.com/user-attachments/assets/b09feb05-551e-44f3-a108-97b4b0c816f8)

as you can see the error in the output SS. the percentage inside the red circle should have been inside the next cell & there should not be any weird < td>.

**so execute the code i have provided in my previous comment. so that you find out the bug yourself. which is surely happening for MIMEText**

**Comment by picnixz** [comment]
> so execute the code i have provided in my previous comment. so that you find out the bug yourself

Please, it is your responsibility to provide a clean and precise description of what is happening. I've reformatted your issue since it's impossible for a reader to read the HTML like that (and not everyone wants to just run arbitrary code).

---

Now, I've used your code (but instead used a local SMTP server instead of gmail). Note that I observed that I need to split your long HTML into lines, otherwise I get an error (but that is likely the issue on the local server side). This may be related but [RFC 5321, §4.5.3.1.6](https://datatracker.ietf.org/doc/html/rfc5321#section-4.5.3.1.6) states that the line length should not exceed 1000 octets (unless extensions are used). And behold!

```py
>>> html = """<html lang='en'><head><meta charset='UTF-8'></head><body><table border=1 bordercolor='black' cellspacing=0 cellpadding=5> <tr> <th>GroupName</th> <th>DS_ID</th> <th>DS_Name</th> <th>DS_MedExpectedSize</th> <th>DS_CurrentSize</th> <th>DS_PercentIncDec</th> <th>TG_ID</th> <th>TG_Name</th> <th>TG_MedExpectedSize</th> <th>TG_CurrentSize</th> <th>TG_PercentIncDec</th> <th>TG_Distinct_MedExpectedSize</th> <th>TG_Distinct_CurrentSize</th> <th>TG_Distinct_PercentIncDec</th></tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2453</td>  <td>TG - Country distribution - Operations</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  <td>2396</td>  <td>3237</td>  <td>35%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2585</td>  <td>TG - Country distribution - Perfumery</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  <td>1289</td>  <td>2983</td>  <td>131%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td>2587</td>  <td>TG - Country distribution - TTH</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  <td>1094</td>  <td>152</td>  <td>-86%</td>  </tr>   <tr> <td>DSM</td>  <td>222</td>  <td>DSM users 2 - Department distribution</td>  <td>0</td>  <td>142</td>  <td>N/A</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>EZ</td>  <td>353</td>  <td>EZK users</td>  <td>18</td>  <td>-3</td>  <td>-117%</td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  <td></td>  </tr>   <tr> <td>KPN</td>  <td>696</td>  <td>KPN HR Servicedesk tickets</td>  <td>123</td>  <td>64</td>  <td>-48%</td>  <td>2512</td>  <td>KPN HR Servicedesk tickets</td>  <td>111</td>  <td>60</td>  <td>-46%</td>  <td>96</td>  <td>40</td>  <td>-58%</td>  </tr>   <tr> <td>NS</td>  <td></td><td></td>  <td></td>  <td></td>  <td></td>  <td>2291</td>  <td>NS TG SSR - Email as Invitation Key - Prio 2</td>  <td>485</td>  <td>676</td>  <td>39%</td>  <td>258</td>  <td>408</td>  <td>58%</td>  </tr></table>    <br><br>   <div>Average expected DS/TG size on a Tuesday, will be calculated as the average TG sizes of the past 10 Tuesday.</div>   </body></html>"""
>>> len(html)
2357
>>> html[1000:1100]
'>2983</td>  <td>131%</td>  </tr>   <tr> <td>DSM</td>  <td>679</td>  <td>DSM users 2 - Country distri'
```

So, this is exactly where you have the issue. I suspect that either us or gmail is actually stopping the HTML at some point (or maybe it's us, I don't know) and the content ends up truncated and ill-formed.

---

Can you try pre-processing your input as follows (or write your HTML line by line instead of a single line):

```py
import textwrap
html = '\n'.join(textwrap.wrap(html, break_on_hyphens=False, break_long_words=False, expand_tabs=False, drop_whitespace=False, replace_whitespace=False))
```

**Comment by azmee-meta** [comment]
i understand what you said & the provided pre-processing solution probably solves the issue (worked for the provided html).

however, for your information, the scenery i have provided not always arise. it sometimes occurs with this type of small html. but, sometime with very large html table (same structure many rows), it renders the html nicely and as expected. and in the production environment i am not using gmail smtp, so the production smtp is also acting the same.

finally, Jazakallah for pointing out the issue. i will see if pre-processing the html content solves the issue in the long run.
Jazakallah again.

**Comment by picnixz** [comment]
In HTML, unless you have a `<pre>` text, you can just add newlines (`\n`) to your content before using it in `MIMEText`. What matters is the content, what matters is how long your content is until a newline (`\n`) is encountered.

I'm however unsure whether we can actually do something on our side. We may accidentally break the HTML content where we don't want to. For instance, if we have a HTML with some `<pre>` tags inside, then any `\n` that we add to break the content would also affect the rendered output which we don't want.

**Comment by azmee-meta** [comment]
I understand. Jazakallah for your assistance. It was a pleasure discussing this with you. Best regards.

**Comment by picnixz** [comment]
You're welcome! I'm categorizing this one as a feature but will still be closing it as `not planned` for now because I don't see a way to make it work without accidentally breaking the content (otherwise, the issue will just remain opened for 10 years). A solution would be to create some helper that wraps the HTML content just in  case, but again its behaviour would be probably heuristic.

If someone encounters a similar issue and stumbles upon this issue, but has a workaround for the standard library, then I'll gladly re-open the issue.
